src/mscorlib/src/System/Diagnostics/Stacktrace.cs

```diff
@@ -664,6 +681,37 @@ private static bool ShowInStackTrace(MethodBase mb)
             return !(mb.IsDefined(typeof(StackTraceHiddenAttribute)) || (mb.DeclaringType?.IsDefined(typeof(StackTraceHiddenAttribute)) ?? false));
         }
 
+        private static bool TryResolveStateMachineMethod(ref MethodBase mb, ref Type declaringType)
+        {
+            Debug.Assert(mb != null);
+            Debug.Assert(declaringType != null);
+
+            Type dt = declaringType.DeclaringType;
+            if (dt == null)
+            {
+                return false;
+            }
+
+            MethodInfo[] methods = dt.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.Instance | BindingFlags.DeclaredOnly);
+            if (methods == null)
+            {
+                return false;
+            }
+
+            foreach (MethodInfo method in methods)
+            {
+                StateMachineAttribute asma = method.GetCustomAttribute<StateMachineAttribute>();
```

Comment:
>method.GetCustomAttribute<StateMachineAttribute>(); [](start = 45, length = 51)

This might need a bit hardening against multiple attributes of the same type applied to the method. It's not supposed to happen, but nothing prevents me to emit metadata like that. I would call something like ```GetCustomAttributes<StateMachineAttribute>().FirstOrDefault();```

Reply:
Only returns a single `StateMachineAttribute`? Will that throw if there is more than one?

Reply:
I think so: http://source.dot.net/#System.Private.CoreLib/src/System/Attribute.cs,662

Reply:
Updated